packages/store/src/decorators/selector.ts

```diff
@@ -1,38 +1,45 @@
+import { CONFIG_MESSAGES, VALIDATION_CODE } from '../configs/messages.config';
 import { createSelector } from '../utils/selector-utils';
 
 /**
  * Decorator for memoizing a state selector.
  */
-export function Selector(selectors?: any[]) {
-  return (target: any, methodName: string, descriptor: PropertyDescriptor) => {
-    if (descriptor.value !== null) {
-      const originalFn = descriptor.value;
-      let memoizedFn: any = null;
-      const newDescriptor = {
-        configurable: true,
-        get() {
-          // Selector initialisation defered to here so that it is at runtime, not decorator parse time
-          memoizedFn =
-            memoizedFn ||
-            createSelector(
-              selectors,
-              originalFn,
-              {
-                containerClass: target,
-                selectorName: methodName,
-                getSelectorOptions() {
-                  return {};
-                }
-              }
-            );
-          return memoizedFn;
-        }
-      };
-      // Add hidden property to descriptor
-      (<any>newDescriptor)['originalFn'] = originalFn;
-      return newDescriptor;
-    } else {
-      throw new Error('Selectors only work on methods');
+export function Selector(selectors?: any[]): MethodDecorator {
+  return <T>(
+    target: Object,
+    propertyKey: string | symbol,
+    descriptor: TypedPropertyDescriptor<T>
+  ): TypedPropertyDescriptor<T> | void => {
+    const isNotMethod: boolean = !(descriptor && descriptor.value !== null);
```

Comment:
Should we add a check that the method is `static`?

Reply:
Not sure

Reply:
@markwhitfeld

Reply:
@splincode as we know that `NGXS_META` is applied to the class, not to the instance, right? So I guess that the `@Selector()` decorator can be used with static methods only. Thoughts?

Reply:
I agree with you

Reply:
We can't use the `NGXS_META` to detect this because it is valid to add selectors to a non-state class. See: https://ngxs.gitbook.io/ngxs/concepts/select#meta-selectors

Also, they could potentially not be static. Nothing is stopping a dev from defining them on an instance method and passing that instance (with its method) around to then be used by store.select(...)

Reply:
done